src/io_callbacks.zig

```diff
@@ -0,0 +1,1189 @@
+                // Wait for some completions and then try again:
+                // See https://github.com/axboe/liburing/issues/281 re: error.SystemResources.
+                // Be careful also that copy_cqes() will flush before entering to wait (it does):
+                // https://github.com/axboe/liburing/commit/35c199c48dfd54ad46b96e386882e7ac341314c5
+                error.CompletionQueueOvercommitted, error.SystemResources => {
+                    try self.flush_completions(true);
+                    continue;
+                },
+                else => return err,
+            };
+            self.submitted += self.queued;
+            self.queued = 0;
+            break;
+        }
+    }
+
+    fn enqueue(self: *IO, completion: *Completion) void {
+        const sqe = self.ring.get_sqe() catch |err| switch (err) {
+            error.SubmissionQueueFull => {
+                self.unqueued.push(completion);
+                return;
+            },
+        };
+        self.queued += 1;
+        completion.prep(sqe);
+    }
+
+    /// This struct holds the data needed for a single io_uring operation
+    pub const Completion = struct {
+        io: *IO,
+        result: i32 = undefined,
+        next: ?*Completion = null,
+        operation: Operation,
+        // This is one of the usecases for c_void outside of C code and as such c_void will
+        // be replaced with anyopaque eventually: https://github.com/ziglang/zig/issues/323
+        context: ?*c_void,
+        callback: fn (context: ?*c_void, completion: *Completion, result: *const c_void) void,
+
+        fn prep(completion: *Completion, sqe: *io_uring_sqe) void {
+            switch (completion.operation) {
+                .accept => |*op| {
+                    linux.io_uring_prep_accept(sqe, op.socket, &op.address, &op.address_size, op.flags);
+                },
+                .close => |op| {
+                    linux.io_uring_prep_close(sqe, op.fd);
+                },
+                .connect => |*op| {
+                    linux.io_uring_prep_connect(sqe, op.socket, &op.address, op.address_size);
+                },
+                .fsync => |op| {
+                    linux.io_uring_prep_fsync(sqe, op.fd, op.flags);
+                },
+                .openat => |op| {
+                    linux.io_uring_prep_openat(sqe, op.fd, op.path, op.flags, op.mode);
+                },
+                .read => |op| {
+                    linux.io_uring_prep_read(sqe, op.fd, op.buffer[0..buffer_limit(op.buffer.len)], op.offset);
+                },
+                .recv => |op| {
+                    linux.io_uring_prep_recv(sqe, op.socket, op.buffer, op.flags);
+                },
+                .send => |op| {
+                    linux.io_uring_prep_send(sqe, op.socket, op.buffer, op.flags);
+                },
+                .timeout => |op| {
+                    // TODO Use proper 64 bit timespec type: https://github.com/ziglang/zig/pull/8118
```

Comment:
We are tracking master so we could do this one now.

Reply:
Sure, I been using a rather out of date master commit so that `vr.zig` still builds, the `std.fmt` changes to require `{s}` for printing slices as a string broke things.

Reply:
Ah, yes I thought you would have spotted something like that since I hadn't updated for awhile. I've updated all the format specifiers in `vr.zig` so that it now builds on `zig-0.8.0-dev.1471+4e9894cfc`.